File: js/popnei/src/pca.ts
/**
 * The principal component analysis of a table that the page holds, and of
 * the variants of a dataset.
 *
 * A principal component analysis places the rows of a table, the
 * individuals, on a few axes that hold as much of the variation between them
 * as that many axes can. Each column of the table, a trait, is centered, its
 * mean taken from it, and standardized, divided by its standard deviation,
 * and the components are the directions in the space of the traits along
 * which the rows vary most: the first has the largest variance that any
 * direction has, the second the largest among the directions at a right
 * angle to the first, and so on.
 *
 * `doPca` takes the table the user brings, and `doPcaFromVariants` makes the
 * table out of the variants of a dataset: each variant becomes one number
 * per individual, its dosage, how many alleles of its genotype are not the
 * major allele of that variant. `docs/specs/pca.md` has what each of the two
 * computes. `PcaResult` is what both give, and `VariantsPcaResult` adds the
 * names of the individuals, the variants that were used and the counts of
 * the pass to it.
 */

import {
  default_center_data as defaultCenterData,
  default_num_prin_comps as defaultNumPrinComps,
  default_standardize_data as defaultStandardizeData,
  default_transform_to_biallelic as defaultTransformToBiallelic,
  pca as pcaOfTheCore,
} from "../wasm/popnei.js";

import {
  aBoolean,
  valuesOfATable,
  wholeNumberOfOneOrMore,
  wholeNumberOfZeroOrMore,
} from "./arguments.js";
import { theWasmHasToBeLoaded } from "./core.js";
import type { PassStats, Variants } from "./variant.js";
import { passStatsOf, sourceOfTheVariants } from "./variant.js";

/**
 * What a principal component analysis gives.
 *
 * Only the components that have variance are here: centering takes one
 * dimension out of the data, so a table of 8 rows and 30 traits has 7
 * components and not 8. In each component the projection of the largest
 * absolute value is positive, which is the rule that makes the numbers the
 * same whichever library did the decomposition, in TypeScript as in Python,
 * and the weights of that component have the sign that rule gave it.
 *
 * The names of the rows and of the traits stay with the application: the
 * rows of `projections` are the rows of the table in the order they were
 * given, and the columns of `princomps` are its traits in their order.
 */
export interface PcaResult {
  /** How many components have variance, which is how many are given. */
  readonly numComps: number;
  /**
   * Where each row falls along each component, the rows of the table x
   * `numComps`, row after row.
   */
  readonly projections: Float64Array;
  /**
   * The variance each component holds as a percentage of the variance of
   * every component of the data, the ones with no variance counted, one
   * number per component.
   */
  readonly explainedVariancePercent: Float64Array;
  /**
   * How many components the weights are given for, which for a table is
   * `numComps`.
   */
  readonly numPrinComps: number;
  /**
   * The weight of each trait in each component, `numPrinComps` x the traits
   * of the table, row after row.
   */
  readonly princomps: Float64Array;
}

/** The two steps on the traits of a table before its components are taken. */
export interface DoPcaOptions {
  /**
   * Whether the mean of each trait is taken from it. True when it is not
   * given. Without it the first component mostly points at the mean of the
   * data.
   */
  centerData?: boolean;
  /**
   * Whether each trait is then divided by its standard deviation, which puts
   * traits measured in different units on one scale; without it the traits
   * with the largest numbers dominate. True when it is not given, and true
   * with `centerData` false is an `Error`: the standard deviation it divides
   * by is the one the trait has once it is centered.
   *
   * The divisor of that standard deviation is the number of rows and not the
   * number of rows less one, which is pyNei's. R's `prcomp` divides by the
   * number of rows less one, so the projections it gives for a standardized
   * table are these times the square root of (n - 1) / n, 0.9975 at 200
   * rows; the percentages and the weights are the same in both, and so are
   * the projections of a table that is not standardized.
   */
  standardizeData?: boolean;
}

/**
 * The principal components of `data`, a table of `numRows` rows of `numCols`
 * values each, row after row, the rows being the individuals and the columns
 * the traits.
 *
 * Every trait is a column of the weights, the one with no variance included,
 * which gets a weight of 0 when the table is not standardized. No value may
 * be missing: a table comes whole.
 *
 * It is pyNei's `do_pca` and popnei's `do_pca` in Python, which take the
 * table as a pandas frame and put the names of its index and of its columns
 * on the result; here the two sides of the table are given as numbers and
 * the names stay with the application.
 *
 * A table whose smaller side is more than 9381 is an `Error` here, and in
 * Python the same table is analysed. What is decomposed is the square of
 * that side, and it, its eigenvectors and the workspace of the
 * eigendecomposition are about 6 times 8 bytes per pair of it, which at 9382
 * is more than the 4 GB a page holds at a time. It is a limit of the browser
 * and not of popnei: WebAssembly addresses 4 GB in one page, and an
 * allocation that does not fit ends the module where an `Error` belongs.
 *
 * @throws {Error} When `data` is not a `Float64Array` of `numRows` times
 * `numCols` values, when a side of the table is not a whole number of 1 or
 * more, when an option is not a boolean, when the smaller side of the table
 * is more than 9381, when a value of the table is not finite, when the table
 * is to be standardized and not centered, when it has fewer than 2 rows or
 * no traits, when it is standardized and a trait has no variance, when no
 * trait of it has any, when the values of a trait are so large or so small
 * that its mean or its standard deviation is not a number the analysis can
 * use, when the linear algebra of the analysis could not be done, and when
 * `init` has not been awaited.
 */
export function doPca(
  data: Float64Array,
  numRows: number,
  numCols: number,
  options: DoPcaOptions = {},
): PcaResult {
  theWasmHasToBeLoaded();
  const rows = wholeNumberOfOneOrMore("numRows", numRows);
  const cols = wholeNumberOfOneOrMore("numCols", numCols);
  const values = valuesOfATable("data", data, rows, cols);
  // The two defaults are the core's, as the ploidy of `openVcf` is, so that
  // Python and TypeScript cannot drift apart on what a table is analysed as
  // when the caller says nothing.
  const centerData =
    options.centerData === undefined
      ? defaultCenterData()
      : aBoolean("centerData", options.centerData);
  const standardizeData =
    options.standardizeData === undefined
      ? defaultStandardizeData()
      : aBoolean("standardizeData", options.standardizeData);
  const result = pcaOfTheCore(values, rows, cols, centerData, standardizeData);
  try {
    return {
      numComps: result.num_comps(),
      projections: theValuesOf(result.projections(), "projections"),
      explainedVariancePercent: theValuesOf(
        result.explained_variance_percent(),
        "explainedVariancePercent",
      ),
      numPrinComps: result.num_prin_comps(),
      princomps: theValuesOf(result.princomps(), "princomps"),
    };
  } finally {
    result.free();
  }
}

/**
 * What the principal components of the variants of a dataset give, which is
 * what a table's analysis gives with the names of the individuals, the
 * variants that were used and the counts of the pass beside it.
 */
export interface VariantsPcaResult extends PcaResult {
  /**
   * The names of the individuals, in the order the source has them, which
   * is the order of the rows of `projections`. They are the names of the
   * rows of the table that a user of `doPca` keeps for themselves: here the
   * source carries them.
   */
  readonly individuals: readonly string[];
  /**
   * Where each individual falls along each component, the individuals x
   * `numComps`, row after row. Every component that has variance is here,
   * whatever `numPrinComps` was.
   */
  readonly projections: Float64Array;
  /**
   * How many components the weights are given for: `numPrinComps` of the
   * call, or the components that have variance when fewer were found, or 0
   * when no weights were asked for.
   */
  readonly numPrinComps: number;
  /**
   * The weight of each variant that was used in each component,
   * `numPrinComps` x `usedVars.length`, row after row. It has no row when no
   * weights were asked for, and `usedVars` names its columns then as well.
   */
  readonly princomps: Float64Array;
  /**
   * The position of each variant that was used, from 0, among the variants
   * the pass gave, the ones with no variance included. They are the columns
   * of `princomps`, and they are given whether or not any weights were
   * asked for.
   */
  readonly usedVars: Uint32Array;
  /**
   * How many variants the pass gave, used or not, and what each filter of
   * the `Variants` was given and kept. The analysis reads the source twice
   * when it is asked for weights, both passes count the same, and these are
   * the counts of the first.
   */
  readonly passStats: PassStats;
}

/** How the principal components of the variants of a dataset are taken. */
export interface DoPcaFromVariantsOptions {
  /**
   * Whether every allele that is not the major one counts the same, which
   * is what gives a variant of more than two alleles a dosage. False when
   * it is not given, and such a variant is then an `Error`.
   */
  transformToBiallelic?: boolean;
  /**
   * How many components the weight of each variant is given for, a whole
   * number of 0 or more. 10 when it is not given, more than the components
   * there are gives those there are, and 0 gives no weight and reads the
   * source once instead of twice.
   *
   * It cuts the weights and nothing else: `projections` and
   * `explainedVariancePercent` are of every component that has variance
   * whatever it is. What it is for is the weights of a million variants,
   * which are 80 MB for 10 components and 8 GB for every component of a
   * dataset of 1000 individuals.
   */
  numPrinComps?: number;
}

/**
 * The principal components of the variants of `variants`, after its steps.
 *
 * Each variant becomes one number per individual, its dosage: how many
 * alleles of the genotype are not the major allele of that variant, which is
 * the most frequent among its called alleles. A genotype with an allele
 * missing takes the mean of the dosages of its variant, so that after
 * centering it pulls its individual nowhere. A variant whose called
 * genotypes all have one dosage has no variance and is left out, a variant
 * with one allele and one where every individual is heterozygous among them;
 * `usedVars` is the ones that were used.
 *
 * The source is read once for the components and a second time for the
 * weights, since a weight needs the eigenvectors and those are known when
 * the first reading ends. Nothing of the size of the variants x the
 * individuals is held: what stays between two blocks is the individuals x
 * individuals matrix. Both readings go through the steps the `Variants` has
 * when the call starts.
 *
 * It is pyNei's `do_pca_from_variants`, whose filters are steps of the
 * `Variants` here, and which gives the weights of every variant where this
 * gives those of the first `numPrinComps` components.
 *
 * A page holds 4 GB at a time, and the matrix of the individuals, its
 * eigenvectors and the workspace of the eigendecomposition are about 6 times
 * 8 bytes per pair of individuals, so a dataset of more than 9381
 * individuals is an `Error` here and is analysed by a program outside the
 * browser, popnei in Python among them.
 *
 * @throws {Error} When `variants` is not a `Variants` or was freed, when
 * `transformToBiallelic` is not a boolean, when `numPrinComps` is not a
 * whole number of 0 or more and at most 4294967295, when the analysis of
 * these individuals does not fit in the memory of a page, when the source
 * cannot be read, a wrong line of a VCF among the causes, when a variant has
 * more than two alleles among its called genotypes and `transformToBiallelic`
 * is false, when the pass gives no variant or no variant with variance, when
 * the source has no individual, when the ploidy is above 254, the
 * individuals are more than 46340 or the variants or the weights are more
 * than a whole number of WebAssembly counts, when the linear algebra could
 * not be done, and when `init` has not been awaited.
 */
export function doPcaFromVariants(
  variants: Variants,
  options: DoPcaFromVariantsOptions = {},
): VariantsPcaResult {
  theWasmHasToBeLoaded();
  const { source, steps } = sourceOfTheVariants("variants", variants);
  const transformToBiallelic =
    options.transformToBiallelic === undefined
      ? defaultTransformToBiallelic()
      : aBoolean("transformToBiallelic", options.transformToBiallelic);
  const numPrinComps =
    options.numPrinComps === undefined
      ? defaultNumPrinComps()
      : wholeNumberOfZeroOrMore("numPrinComps", options.numPrinComps);
  // The steps of the two passes are a copy of the list, made after the
  // arguments were checked so that nothing refused here leaves one behind:
  // the call takes it over and frees it.
  const result = source.pca_of_variants(
    transformToBiallelic,
    numPrinComps,
    steps.of_a_pass(),
  );
  try {
    return {
      individuals: variants.individuals,
      numComps: result.num_comps(),
      projections: theValuesOf(result.projections(), "projections"),
      explainedVariancePercent: theValuesOf(
        result.explained_variance_percent(),
        "explainedVariancePercent",
      ),
      numPrinComps: result.num_prin_comps(),
      princomps: theValuesOf(result.princomps(), "princomps"),
      usedVars: thePositionsOf(result.used_vars(), "usedVars"),
      passStats: passStatsOf(result.pass_stats()),
    };
  } finally {
    result.free();
  }
}

/**
 * The positions of one array of the result.
 *
 * It is `theValuesOf` for the array of positions, which is a
 * `Uint32Array`: each array of a result leaves the memory of wasm the first
 * time it is asked for.
 *
 * @throws {Error} When the array had been read already.
 */
function thePositionsOf(
  positions: Uint32Array | undefined,
  name: string,
): Uint32Array {
  if (positions === undefined) {
    throw new Error(
      `popnei: \`${name}\` was read twice out of the memory of WebAssembly, ` +
        "which is a defect of popnei; please report it",
    );
  }
  return positions;
}

/**
 * The values of one array of the result.
 *
 * Each array leaves the memory of wasm the first time it is asked for, so
 * the call after that gives nothing; `doPca` and `doPcaFromVariants` read
 * each of them once, and the `Error` here is a defect of this package and
 * not something a caller can do.
 *
 * @throws {Error} When the array had been read already.
 */
export function theValuesOf(
  values: Float64Array | undefined,
  name: string,
): Float64Array {
  if (values === undefined) {
    throw new Error(
      `popnei: \`${name}\` was read twice out of the memory of WebAssembly, ` +
        "which is a defect of popnei; please report it",
    );
  }
  return values;
}

File: python/popnei/pca.py
"""The principal component analysis: where the individuals fall on a few axes.

A principal component analysis places the individuals of a dataset on a few
axes that hold as much of the variation between them as that many axes can,
which is how a user sees whether their individuals fall into populations
before they name any. :func:`do_pca` does it on a table of numbers that the
user brings, individuals x traits, and :func:`do_pca_from_variants` on the
variants of a :class:`popnei.Variants`, where each variant is a trait and
the number of an individual at it is its dosage.

`docs/specs/pca.md` has what is computed. Each trait is centered, its mean
taken from it, and standardized, divided by its standard deviation, which
puts traits measured in different units on one scale; the components are
then the directions in the space of the traits along which the individuals
vary most, the first holding the largest variance that any direction has,
the second the largest among the directions at a right angle to the first,
and so on.
"""

from dataclasses import dataclass

import numpy
import pandas

from popnei import _core
from popnei.variant import PassStats, Variants, _pass_stats_of

# What is wrong with a trait whose mean or whose standard deviation the
# analysis cannot use, under the name the binding crate gives each of the
# three. The values of such a trait are too large or too small for the
# arithmetic of a float64, and what the user does about it is to scale that
# trait or take it out.
_WHAT_IS_WRONG_WITH_THE_TRAIT = {
    "mean_not_finite": (
        "its values sum above the largest float64, 1.8e308, so its mean is not "
        "finite and every value of it would become a NaN"
    ),
    "deviation_not_finite": (
        "the squares of its deviations sum above the largest float64, so its "
        "standard deviation is not finite and the trait would become a column "
        "of zeros"
    ),
    "deviation_of_zero": (
        "the squares of its deviations all fall below the smallest float64 "
        "above 0, 5e-324, so its standard deviation is 0 although its values "
        "differ, and dividing by it would give infinities"
    ),
}


@dataclass(frozen=True)
class PCAResult:
    """What a principal component analysis gives.

    Only the components that have variance are here: centering takes one
    dimension out of the data, so a table of 8 individuals and 30 traits has
    7 components and not 8. They are named ``PC0``, ``PC1`` and so on, with
    zeros on the left to the width of how many there are, ``PC000`` for 200
    of them, and the same names are the columns of `projections`, the index
    of `explained_variance_percent` and the index of `princomps`.

    A component multiplied by -1 is the same component, and which of the two
    a decomposition gives depends on the library that did it. popnei fixes
    it: in each component the projection of the largest absolute value is
    positive, and the weights of that component have the sign that gave it.
    So the numbers are the same in Python, under pyodide and in TypeScript.
    """

    projections: pandas.DataFrame
    """Where each individual falls along each component, one row per
    individual and one column per component. The index is the index of the
    table, or the names of the individuals of the ``Variants``."""

    explained_variance_percent: pandas.Series
    """How much of the variance each component holds, as a percentage of the
    variance of every component there is, given or not."""

    princomps: pandas.DataFrame
    """The weight of each trait in each component, one row per component and
    one column per trait. The columns are the columns of the table, or the
    position of each variant that was used among the variants the pass gave,
    from 0. Of a table there is one row per component; of the variants there
    are ``num_prin_comps``, which can be fewer, and they are the first
    components."""

    pass_stats: PassStats | None
    """The counts of the pass over the source of variants, and ``None`` for
    the analysis of a table, which reads no variants."""


def do_pca(
    data: pandas.DataFrame,
    center_data: bool = _core.DEFAULT_CENTER_DATA,
    standardize_data: bool = _core.DEFAULT_STANDARDIZE_DATA,
) -> PCAResult:
    """The principal components of `data`, a table of individuals x traits.

    The index of `data` names the rows of the projections of the result and
    its columns name the columns of the weights. Every component that has
    variance is given, with its weights: the table is in memory already, so
    there is nothing to ask for fewer of. The result reads no variants, so
    its ``pass_stats`` is ``None``.

    `center_data` takes the mean of each trait from it. Without it the first
    component mostly points at the mean of the data.

    `standardize_data` then divides each trait by its standard deviation,
    which puts traits measured in different units on one scale; without it
    the traits with the largest numbers dominate. The divisor of that
    deviation is the number of individuals and not the number less one,
    which is pyNei's and makes every projection of a standardized table
    0.9975 of what R's ``prcomp`` gives at 200 individuals. Standardizing
    divides by the deviation the trait has once it is centered, so asking
    for it with `center_data` false is a ``ValueError``.

    No value may be missing: a table comes whole. A value that is not
    finite, an infinity or a NaN, and one that pandas holds as missing in a
    nullable dtype, are a ``ValueError`` that says which row and which trait
    it is at. So are a table of fewer than 2 rows or of no traits, and one
    in which no trait has variance once it is centered.

    Two more ``ValueError`` name the trait they are about, as the frame
    names it. One is a trait with no variance, every value of it equal to
    the others, when the table is standardized: there is nothing to divide
    it by, and the message says how many such traits there are and names the
    first ten of them, so that the user can take them out. Without
    standardizing such a trait is no error and gets a weight of 0. The other
    is a trait whose values are too large or too small for the arithmetic of
    a float64, whose message says which of those two it is.

    It is pyNei's ``do_pca``, which spells `standardize_data`
    ``standarize_data``, gives the components that have no variance as well,
    and leaves the sign of each component to the library that decomposed the
    table.
    """
    values = numpy.ascontiguousarray(
        # A frame of a nullable dtype holds its missing values as pandas's
        # own NA, which numpy cannot turn into a float64 on its own. It
        # arrives as a NaN, and the core says which row and which trait it
        # is at, instead of numpy raising a TypeError about a dtype.
        data.to_numpy(dtype=numpy.float64, na_value=numpy.nan),
        dtype=numpy.float64,
    )
    try:
        projections, percent, princomps = _core.pca(
            values, center_data, standardize_data
        )
    except _core.TraitsWithNoVariance as error:
        raise ValueError(
            _the_traits_with_no_variance(error.args[1], data.columns)
        ) from None
    except _core.TraitOutOfRange as error:
        raise ValueError(
            _the_trait_out_of_range(error.args[1], error.args[2], data.columns, error)
        ) from None
    names = _component_names(projections.shape[1])
    return PCAResult(
        projections=pandas.DataFrame(projections, index=data.index, columns=names),
        explained_variance_percent=pandas.Series(percent, index=names),
        princomps=pandas.DataFrame(princomps, index=names, columns=data.columns),
        pass_stats=None,
    )


def do_pca_from_variants(
    variants: Variants,
    transform_to_biallelic: bool = _core.DEFAULT_TRANSFORM_TO_BIALLELIC,
    num_prin_comps: int = _core.DEFAULT_NUM_PRIN_COMPS,
) -> PCAResult:
    """The principal components of the variants of `variants`.

    Each variant becomes one number per individual, its dosage: how many
    alleles of the genotype are not the major allele of the variant, which
    is the most frequent among its called alleles and the lowest numbered of
    two that are equally frequent. A genotype with an allele missing takes
    the mean of the dosages of its variant, so that after centering it pulls
    its individual nowhere, and the standard deviation each variant is
    divided by has all the individuals in it and not the called ones alone.

    A variant whose called genotypes all have one dosage has no variance and
    is left out, a variant with one allele and one where every individual is
    heterozygous among them, and so is a variant with no called genotype.
    The ones that were used are the columns of the weights, by their
    position among the variants the pass gave, from 0.

    The names of the individuals are the index of the projections, and the
    counts of the pass are in ``pass_stats``: ``num_vars`` is how many
    variants the steps of the `Variants` let through, used or not, and
    ``filtering`` what each filter of the pass was given and kept.

    `transform_to_biallelic` makes every allele that is not the major one
    count the same, which is what a variant of more than two different
    alleles among its called genotypes needs: the dosage of a genotype has a
    meaning for two alleles, and without this such a variant is a
    ``ValueError`` that says which one it is. The alleles are those the
    genotypes hold and not those the source lists.

    `num_prin_comps` is how many components the weights are given for, 10 by
    default, and it is new here: pyNei gives the weight of every variant in
    every component, which is 0.8 GB for 100000 variants of 1000
    individuals. They come from a second pass over the variants, because a
    weight needs the eigenvectors, which are known when the first pass ends,
    so with 0 there is no second pass, ``princomps`` has no rows and it
    still has the variants that were used as its columns. More components
    than there are gives those there are, and a number below 0 is a
    ``ValueError``, as is one above what the machine counts; what is no
    whole number, a float, a string and a truth value among them, is a
    ``TypeError`` that names the argument and what was given. The
    projections and the percentages are of every component that has variance
    whatever `num_prin_comps` is.

    A dataset with no variant, one where no variant has variance, which one
    individual gives, and one with no individual are a ``ValueError``. So is
    a dataset of a size this analysis cannot count in, which is one of four:
    a ploidy above 254, more than 46340 individuals, more variants than the
    machine counts, which under pyodide is 4295 million, and weights of more
    values than the machine counts, which is `num_prin_comps` times the
    variants that were used and which fewer variants reach the more
    components are asked for.

    It is pyNei's ``do_pca_from_variants``, with `num_prin_comps` added,
    without `num_threads`, which no calculation of popnei takes, and with
    the filters of the ``Variants`` as steps of it and not as functions
    around it. pyNei counts the alleles of a whole chunk of variants and
    popnei those of each variant; pyNei gives the components with no
    variance as well; and a variant with no called genotype is an error in
    pyNei and is left out here.
    """
    if not isinstance(variants, Variants):
        # What a user gives instead is usually the path of the VCF, and
        # what that gave was the `AttributeError` of an object with no
        # source inside it.
        raise TypeError(
            f"`variants` is {variants!r}, a {type(variants).__name__}, and "
            f"`do_pca_from_variants` reads the variants of a source: give it "
            f"what `open_vcf` or `open_vars` gives, "
            f"do_pca_from_variants(open_vcf(vcf_path))"
        )
    # `num_prin_comps` is checked in the binding crate, where every count a
    # user writes is: a whole number of Python is of any size, and what is
    # none of them is refused there by the name of the argument.
    projections, percent, princomps, used_vars, counts = _core.pca_of_variants(
        variants._source, transform_to_biallelic, num_prin_comps, variants._steps
    )
    # The components of the projections are every one that has variance, and
    # the weights are those of the first `num_prin_comps` of them, so the
    # names are made for the projections and the weights take the first of
    # them: one component has one name in both frames.
    names = _component_names(projections.shape[1])
    return PCAResult(
        projections=pandas.DataFrame(
            projections, index=list(variants.individuals), columns=names
        ),
        explained_variance_percent=pandas.Series(percent, index=names),
        princomps=pandas.DataFrame(
            princomps, index=names[: princomps.shape[0]], columns=used_vars
        ),
        pass_stats=_pass_stats_of(counts),
    )


def _component_names(num_comps: int) -> list[str]:
    """`PC0`, `PC1` and so on, one for each component.

    The number has zeros on its left to the width of `num_comps`, ``PC000``
    for 200 components, which is what pyNei's `_create_pc_names` gives and
    what makes the names sort as their numbers do.
    """
    width = len(str(num_comps))
    return [f"PC{number:0{width}d}" for number in range(num_comps)]


def _the_traits_with_no_variance(
    positions: list[int], trait_names: pandas.Index
) -> str:
    """What a user is told of the traits that cannot be standardized.

    They are named as the columns of their frame, the first ten of them and
    how many more there are, so that a user of a table of hundreds of traits
    reads a message of one line and knows how many to take out. The core has
    said the same with the position of each trait in the place of its name.
    """
    shown = ", ".join(f"`{trait_names[position]}`" for position in positions[:10])
    left_out = len(positions) - len(positions[:10])
    more = "" if left_out == 0 else f", and {left_out} more"
    return (
        f"{len(positions)} of the {len(trait_names)} traits have no variance and "
        f"cannot be standardized: take them out of the table or do not standardize; "
        f"they are the traits {shown}{more}"
    )


def _the_trait_out_of_range(
    position: int,
    problem: str,
    trait_names: pandas.Index,
    of_the_core: BaseException,
) -> str:
    """What a user is told of a trait the analysis cannot scale.

    The trait is named as its column of the frame is, and the message says
    which of the three things happened to it, so that the user knows whether
    to scale it or to take it out. A problem this package has no words for,
    which a core newer than it can give, is passed on as the core said it,
    with the position of the trait in the place of its name.
    """
    said = _WHAT_IS_WRONG_WITH_THE_TRAIT.get(problem)
    if said is None:
        return str(of_the_core.args[0])
    return (
        f"the trait `{trait_names[position]}` cannot be centered or standardized: "
        f"{said}; scale that trait or take it out of the table"
    )
